airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -110,13 +110,7 @@ def observe(self, new_state: Mapping[str, Any]) -> None:
         pass
 
     def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
-        # This is used to avoid sending a duplicate state message at the end of a sync since the stream has already
-        # emitted state at the end of each slice. We only emit state if _current_slice is None which indicates we had no
-        # slices and emitted no record or are currently in the process of emitting records.
-        if self.current_slice is None or not self._finished_sync:
-            return self._cursor.get_stream_state()
-        else:
-            return None
+        return self._cursor.get_stream_state()
```

Comment:
Wanted to discuss the issue you brought up @tolik0 around missing the last stream state on the PR itself so context isn't lost.

So if I'm understanding the issue, because we are emitting parent slices without waiting, by the time we update `self._parent_state` to its final value at the end of `SubstreamPartitionRouter.stream_slices(), all records and slices were already processed. However, we have not emitted the final parent state. And the checkpoint reader is marked as `_finished_sync = True`, so we don't emit the final state?

If so are the options

1. How you have it now and accept that we emit at least one extra state message per-stream?
2. Modify the internal `PerPartitionCursor` implementation to somehow emit final state. I haven't thought about how to do this, but is this your thinking?
3. Add conditional branch that if this is a `PerPartitionCursor` then we `return self._cursor.get_stream_state()`

I think I prefer option 1 over option 3 because bleeding low-code implementation details is something I'd like avoid if possible. 2 feels like the best option if this doesn't prove to be too difficult. But if it proves very hard, I feel like option 1 is not too bad.

Reply:
Yes, you’re correct. We’re not emitting the final state because we only get the final state after `stream_slices()` finishes. I don’t see how we can implement the second option, as we don’t have access to the stream, and we can’t emit the state directly from `PerPartitionCursor`. I wasn’t sure why we had this workaround, so I was considering options 1 and 3. However, I agree that option 1 is much better.

Reply:
okay cool we are on the same page. My only worry is that we might have some unit/mock server tests break because we might be checking for duplicates or an exact number of state message counts. But updating those seems fine and how you have it (basically option 1) looks fine to me.

Thanks for confirming